src/arch/riscv/semihosting.cc

```diff
@@ -0,0 +1,191 @@
+        {SYS_HEAPINFO, {"SYS_HEAPINFO", &RiscvSemihosting::callHeapInfo32,
+                               &RiscvSemihosting::callHeapInfo64}},
+
+        {SYS_EXIT, {"SYS_EXIT", &RiscvSemihosting::callExit32,
+                           &RiscvSemihosting::callExit64}},
+        {SYS_EXIT_EXTENDED,
+                {"SYS_EXIT_EXTENDED", &RiscvSemihosting::callExitExtended}},
+
+        {SYS_ELAPSED, {"SYS_ELAPSED", &RiscvSemihosting::callElapsed32,
+                              &RiscvSemihosting::callElapsed64}},
+        {SYS_TICKFREQ, {"SYS_TICKFREQ", &RiscvSemihosting::callTickFreq}},
+};
+
+RiscvSemihosting::
+RiscvSemihosting(const RiscvSemihostingParams &p) : BaseSemihosting(p)
+{}
+
+bool
+RiscvSemihosting::call64(ThreadContext *tc)
+{
+    RegVal op = tc->getReg(RiscvISA::int_reg::A0) & mask(32);
+    auto it = calls.find(op);
+    if (it == calls.end()) {
+        unrecognizedCall<Abi64>(tc, "Unknown semihosting call: op = 0x%x", op);
+        return false;
+    }
+    const SemiCall &call = it->second;
+
+    DPRINTF(Semihosting, "Semihosting call64: %s\n", call.dump64(tc));
+    auto err = call.call64(this, tc);
+    semiErrno = err.second;
+    DPRINTF(Semihosting, "\t ->: 0x%x, %i\n", err.first, err.second);
+
+    return true;
+}
+
+bool
+RiscvSemihosting::call32(ThreadContext *tc)
+{
+    RegVal op = tc->getReg(RiscvISA::int_reg::A0);
+    auto it = calls.find(op);
+    if (it == calls.end()) {
+        unrecognizedCall<Abi32>(
+                tc, "Unknown aarch32 semihosting call: op = 0x%x", op);
+        return false;
+    }
+    const SemiCall &call = it->second;
+
+    DPRINTF(Semihosting, "Semihosting call32: %s\n", call.dump32(tc));
+    auto err = call.call32(this, tc);
+    semiErrno = err.second;
+    DPRINTF(Semihosting, "\t ->: 0x%x, %i\n", err.first, err.second);
+
+    return true;
+}
+
+bool
+RiscvSemihosting::call(ThreadContext *tc)
+{
+    auto isa = dynamic_cast<RiscvISA::ISA *>(tc->getIsaPtr());
+    panic_if(!isa, "Cannot derive rv_type from non-riscv isa");
+    return isa->rvType() == enums::RV32 ? call32(tc) : call64(tc);
+}
+
+PortProxy &
+RiscvSemihosting::portProxyImpl(ThreadContext *tc)
+{
+    static std::unique_ptr<PortProxy> port_proxy([=]() {
+        return FullSystem ? new TranslatingPortProxy(tc) :
+                            new SETranslatingPortProxy(tc);
+    }());
+    return *port_proxy;
+}
+
+bool
+RiscvSemihosting::isSemihostingEBreak(ExecContext *xc)
+{
+    // Check if the surrounding bytes match the semihosting magic sequence.
+    PortProxy& proxy = portProxyImpl(xc->tcBase());
+    Addr PrevInstAddr = xc->pcState().instAddr() - 4;
+    Addr NextInstAddr = xc->pcState().instAddr() + 4;
+    if ((PrevInstAddr & ~Addr(15)) != (NextInstAddr & ~Addr(15))) {
+        DPRINTF(Semihosting,
+                "Ebreak cannot be a semihosting ebreak since previous "
+                "and next instruction are not sufficiently aligned\n");
+        return false;
```

Comment:
The spec doesn't required the sequence of instruction to be in the same instruction block(16 bytes), right?

Reply:
Oh indeed that is too strict, they just need to be on the same page (if paging is enabled).

> If paging is in use in the current mode, this sequence must not cross a page boundary as the semihosting system must be able to check for the semihosting sequence without needing data from potentially missing pages.

Reply:
Ok, I would like to check if the instruction sequence is in the same 4K Bytes page block. It will reduce the strict checking.